spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestDescriptor.kt

```diff
@@ -0,0 +1,84 @@
+package org.spekframework.spek2.junit
+
+import org.junit.platform.engine.TestDescriptor
+import org.junit.platform.engine.TestSource
+import org.junit.platform.engine.TestTag
+import org.junit.platform.engine.UniqueId
+import org.junit.platform.engine.support.descriptor.ClassSource
+import org.spekframework.spek2.runtime.scope.*
+import java.util.*
+
+class SpekTestDescriptor internal constructor(
+        val scope: ScopeImpl,
+        private val factory: SpekTestDescriptorFactory
+) : TestDescriptor {
+
+    companion object {
+        private val ENGINE_ID = UniqueId.forEngine(SpekTestEngine.ID)
+    }
+
+    private val id by lazy { computeId(scope) }
+
+    private fun computeId(scope: ScopeImpl?): UniqueId = if (scope == null) {
+        ENGINE_ID
+    } else {
+        computeId(scope.parent as ScopeImpl?).append("${scope.id.type}", scope.id.name)
+    }
+
+    private var engineDescriptor: SpekEngineDescriptor? = null
+    private val childDescriptors = mutableSetOf<TestDescriptor>()
+
+    override fun getUniqueId() = id
+
+    override fun getDisplayName(): String = scope.path.name
+
+    override fun getType(): TestDescriptor.Type = when (scope) {
+        is GroupScopeImpl -> TestDescriptor.Type.CONTAINER
+        is ActionScopeImpl -> TestDescriptor.Type.CONTAINER_AND_TEST
+        is TestScopeImpl -> TestDescriptor.Type.TEST
+    }
+
+    override fun getSource(): Optional<TestSource> = when (scope.id.type) {
+        ScopeType.CLASS -> Optional.of(ClassSource.from(scope.id.name))
+        ScopeType.SCOPE -> Optional.empty()
+    }
+
+    override fun setParent(parent: TestDescriptor?) {
+        // Called only when adding as a child of the engine's descriptor.
+        // Will be null only if it's not the root scope.
+        if (parent != null && parent is SpekEngineDescriptor) {
+            engineDescriptor = parent
+        }
+    }
+
+    override fun getParent(): Optional<TestDescriptor> {
+        // TODO: replace with just `scope.parent`?
```

Comment:
@raniejade, please take a look at this. Seems like the whole expression can be simplified to `scope.parent as ScopeImpl?`.

Reply:
Go ahead 👍.

Reply:
@raniejade, done, thanks!

FYI — I don’t have rights to merge PRs, so please do it yourself if everything else looks fine.